### PR TITLE
Introduce Runtime Security Agent - RFC Security Agent 

### PR DESCRIPTION
**Pull request by safchain**
### What does this PR do?

This PR implements the Security Agent RFC

### Motivation

This PR introduces the Runtime Security Agent with its first feature: File Integrity Monitoring. Most of the code resides in system-probe and leverages eBPF to detect abnormal behaviors based on rules provided in a configuration file.

### Additional Notes

Most of the code resides in system-probe.

- pkg/security/agent:
Receive events sent by the runtime security agent part of system-probe
and forwards them to Datadog

- pkg/security/api:
Protobuf definition of triggered security alerts

- pkg/security/config:
Definition of the confiuration of the runtime security agent

- pkg/security/ebpf:
C source of eBPF programs required by the runtime security agent

- pkg/security/module:
Runtime security system-probe module

- pkg/security/policy:
Policy - set of rules and macros - loading

- pkg/security/probe:
Core of runtime security agent.
Loads eBPF programs, decode and evaluate events sent from the kernel

- pkg/security/rules:
Rules evaluation

- pkg/security/secl:
Rules parser and code generation for fast evaluation

- pkg/security/tests:
Functional test suite

### Describe your test plan

Functional test suite that will be run automatically in a future CI Job

### CODE REVIEW

**Comment by mx-psi** [comment]
You should probably specify the owner of `pkg/security` on the CODEOWNERS file

File: cmd/system-probe/modules/security_runtime.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+	secmodule "github.com/DataDog/datadog-agent/pkg/security/module"
+)
+
+// SecurityRuntime - Security runtime Factory
+var SecurityRuntime = api.Factory{
+	Name: "security_runtime",
+	Fn: func(cfg *config.AgentConfig) (api.Module, error) {
+		return secmodule.NewModule(cfg)
+	},
+}
```
Comment (leeavital) [inline cmd/system-probe/modules/security_runtime.go:15]: Should this module be attached by the security agent and not the system-probe ?

Comment (lebauce) [reply cmd/system-probe/modules/security_runtime.go:15]: The runtime security agent is made of 2 parts :
- one system-probe module for the eBPF code
- an other attached by the security agent that listens for events from the first part and send events to Datadog

File: pkg/config/config_template.yaml [hunk 2]
```diff
@@ -873,6 +873,41 @@ api_key:
...
+  #
+  # socket: /opt/datadog-agent/run/runtime-security.sock
+
+  ## @param policies - custom object - optional
+  ## Policy names and their policy rules.
+  #
+  # policies:
+    # - name: default-policies
+    #   files:
+    #     - /etc/datadog-agent/security_agent.d/runtime-policies.yaml
```
Comment (xornivore) [inline pkg/config/config_template.yaml:899]: How do we envision configuring this through ENV vars? Should it just be a directory mount with all policies automatically loaded? What is the benefit of having individual files listed here?

File: pkg/security/agent/agent.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+			// Dispatch security event
+			rsa.DispatchEvent(in)
+		}
+	}
+}
+
+// DispatchEvent dispatches a security event message to the subsytems of the runtime security agent
+func (rsa *RuntimeSecurityAgent) DispatchEvent(evt *api.SecurityEventMessage) {
+	// For now simply log to Datadog
+	rsa.logClient.SendSecurityEvent(evt, message.StatusAlert)
```
Comment (xornivore) [inline pkg/security/agent/agent.go:95]: This should use the `event.Reporter` for consistency with compliance features and to conform to the same message format.

File: pkg/security/ebpf/table.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+// Bytes returns the binary representation of a Uint64TableItem
+func (i Uint64TableItem) Bytes() ([]byte, error) {
+	b := make([]byte, 8)
+	ByteOrder.PutUint64(b, uint64(i))
+	return b, nil
+}
+
+// StringTableItem describes an string table key or value
+type StringTableItem struct {
+	str  string
```
Comment (xornivore) [inline pkg/security/ebpf/table.go:101]: Should fields be exported in this exported struct? Or should it be implementing an exported interface and be private to the package?

File: releasenotes/notes/consistent-config-secret-decryption-f24dd89f5c485f88.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# Each section from every releasenote are combined when the
```
Comment (xornivore) [inline releasenotes/notes/consistent-config-secret-decryption-f24dd89f5c485f88.yaml:1]: There should probably be a separate release note for the bulk of the code in this PR. For this issue in this release note specifically - is the rebase to master OK? I see this release note in master already?

File: tasks/security_agent.py [hunk 6]
```diff
@@ -29,6 +29,22 @@
 ]
 
 
+def get_go_env(ctx, go_version):
```
Comment (xornivore) [inline tasks/security_agent.py:32]: Why is this needed?

File: tasks/security_agent.py [hunk 7]
```diff
@@ -102,3 +118,25 @@ def gen_mocks(ctx):
...
+@task
+def functional_tests(
+    ctx, race=False, verbose=False, go_version=None, arch="x64", major_version='7', pattern='', output=''
+):
+    ldflags, gcflags, env = get_build_flags(ctx, arch=arch, major_version=major_version)
+
+    goenv = get_go_env(ctx, go_version)
+    env.update(goenv)
+
+    cmd = 'sudo -E go test -tags functionaltests {output_opt} {verbose_opt} {run_opt} {REPO_PATH}/pkg/security/tests'
```
Comment (xornivore) [inline tasks/security_agent.py:132]: Is this running in CI?

File: cmd/system-probe/main.go [hunk 8]
```diff
@@ -119,11 +120,13 @@ func runAgent(exit <-chan struct{}) {
 		utils.WriteAsJSON(w, stats)
 	})
 
-	err = http.Serve(conn.GetListener(), httpMux)
-	if err != nil {
-		log.Criticalf("Error creating HTTP server: %s", err)
-		cleanupAndExit(1)
-	}
+	go func() {
```
[the remark this answers is not in the document]
Comment (lebauce) [reply cmd/system-probe/main.go:123]: This commit is part of a separate PR that we submitted and that was merged recently (https://github.com/DataDog/datadog-agent/pull/5989/)

File: pkg/security/ebpf/table.go (hunk 4 above)
Comment (lebauce) [reply pkg/security/ebpf/table.go:101]: It is implementing the `TableItem` interface and can be instantiated using `NewStringTableItem`. What would be the value of exporting the fields ?

File: tasks/security_agent.py (hunk 6 above)
Comment (lebauce) [reply tasks/security_agent.py:32]: It is used by the `functional_tests` target. In a previous version, we modified all the targets where the logic was copy/pasted (around 5 times) to use this function. But we decided to keep this commit and submit it later, to make the changes on common parts as small as possible. Does it make sense ?

File: tasks/security_agent.py (hunk 7 above)
Comment (lebauce) [reply tasks/security_agent.py:132]: No, and it definitely should be ! This is a long running task that we have in the eBPF squad. The idea is to compile the functional tests as an executable and run them using Kitchen.

File: releasenotes/notes/consistent-config-secret-decryption-f24dd89f5c485f88.yaml (hunk 5 above)
Comment (lebauce) [reply releasenotes/notes/consistent-config-secret-decryption-f24dd89f5c485f88.yaml:1]: This a part of an other cherry-pick of a merged PR. We'll rebase on master to reduce the noise

File: pkg/security/agent/agent.go (hunk 3 above)
Comment (lebauce) [reply pkg/security/agent/agent.go:95]: Ok. We'll update the PR

File: cmd/security-agent/app/runtime.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+	"github.com/DataDog/datadog-agent/pkg/process/config"
+	secagent "github.com/DataDog/datadog-agent/pkg/security/agent"
+	secmodule "github.com/DataDog/datadog-agent/pkg/security/module"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+var (
+	runtimeCmd = &cobra.Command{
+		Use:   "runtime",
+		Short: "Runtime Agent utility commands",
```
[the remark this answers is not in the document]
Comment (lebauce) [reply cmd/security-agent/app/runtime.go:28]: Updated

File: cmd/security-agent/app/app.go [hunk 10]
```diff
@@ -189,6 +189,11 @@ func start(cmd *cobra.Command, args []string) error {
 	}
 	defer srv.Stop()
 
+	// start runtime security agent
+	if err = startRuntimeSecurity(stopper); err != nil {
```
[the remark this answers is not in the document]
Comment (lebauce) [reply cmd/security-agent/app/app.go]: We just rebased against master, changed the behaviour to "exit the process if neither compliance nor runtime security are enabled and moved the instantiation of the runtime security agent just after the compliance one.

File: pkg/autodiscovery/autoconfig.go [hunk 11]
```diff
@@ -281,26 +281,19 @@ func (ac *AutoConfig) processNewConfig(config integration.Config) []integration.
...
-		for _, config := range resolvedConfigs {
-			config, err := decryptConfig(config)
-			if err != nil {
-				log.Errorf("Dropping conf for %q: %s", config.Name, err.Error())
-				continue
-			}
-			configs = append(configs, config)
-		}
-		return configs
+		return resolvedConfigs
```
[the remark this answers is not in the document]
Comment (lebauce) [reply pkg/autodiscovery/autoconfig.go]: Was part of a cherry pick. Not present anymore after rebase

File: cmd/security-agent/app/runtime_unsupported.go [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+func startRuntimeSecurity(stopper restart.Stopper) error {
+	enabled := coreconfig.Datadog.GetBool("runtime_security_config.enabled")
+	if !enabled {
+		log.Info("Datadog runtime security agent disabled by config")
+		return nil
+	}
+
+	log.Error("Datadog runtime security agent is only supported on Linux")
```
[the remark this answers is not in the document]
Comment (lebauce) [reply cmd/security-agent/app/runtime_unsupported.go:23]: On unsupported platforms with the runtime security agent enabled, we prefer to just log an error instead of failing, for safety

File: pkg/config/config_template.yaml (hunk 2 above)
Comment (safchain) [reply pkg/config/config_template.yaml:899]: As suggested we are going to load all the policy files of a path. We will update this PR

File: pkg/security/ebpf/table.go (hunk 4 above)
Comment (xornivore) [reply pkg/security/ebpf/table.go:101]: Thank you. This is a small nit really, but if you can't initialize a struct outside of this package, then maybe it shouldn't be exported? Are there currently use cases that require this type, e.g. in a `switch` on `.(type)`, mind pointing me at them? If they are only used via the implemented interface then maybe keep this and other types private to the package?

File: pkg/config/config_template.yaml (hunk 2 above)
Comment (xornivore) [reply pkg/config/config_template.yaml:899]: Great, thank you. If you want to keep it consistent with existing code in compliance, then maybe `runtime_security_config.dir` or `runtime_security_config.policies.dir` or similar?

File: releasenotes/notes/consistent-config-secret-decryption-f24dd89f5c485f88.yaml (hunk 5 above)
Comment (xornivore) [reply releasenotes/notes/consistent-config-secret-decryption-f24dd89f5c485f88.yaml]: Great, thank you, could you also please remove the `changelog/no-changelog` label from this PR? There should be a release note for these changes.

**Review by mx-psi — COMMENTED** [review]
You should rebase/merge master so that you have the most recent CI tests including macOS tests and Python linting.

File: cmd/security-agent/app/runtime_unsupported.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (xornivore) [reply cmd/security-agent/app/runtime_unsupported.go:23]: Thanks. To clarify this - if it's not enabled by the user (intentionally), it's not necessarily unsupported?

File: pkg/security/utils/mount.go [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+// A MountInfo is a type that describes the details, options
+// for each mount, parsed from /proc/self/mountinfo.
+// The fields described in each entry of /proc/self/mountinfo
+// is described in the following man page.
+// http://man7.org/linux/man-pages/man5/proc.5.html
+type MountInfo struct {
+	// Unique Id for the mount
+	MountID int `json:"moung_id"`
```
Comment (sunhay) [inline pkg/security/utils/mount.go:27]: `moung_id` -> `mount_id`

File: pkg/security/utils/mount.go [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+	return opts
+}
+
+func hostProc() string {
+	value := os.Getenv("HOST_PROC")
+	if value == "" {
+		value = "/proc"
+	}
+	return value
+}
```
Comment (sunhay) [inline pkg/security/utils/mount.go:154]: We have a number of copies of this function within our codebase - we should consider a single util for accessing `proc/`

https://github.com/DataDog/datadog-agent/search?q=HostProc&unscoped_q=HostProc

File: pkg/security/ebpf/table.go (hunk 4 above)
Comment (safchain) [reply pkg/security/ebpf/table.go:101]: There is a constructor for this `NewStringTableItem`

File: pkg/security/utils/mount.go (hunk 14 above)
Comment (lebauce) [reply pkg/security/utils/mount.go]: Removed

File: pkg/security/utils/mount.go (hunk 13 above)
Comment (lebauce) [reply pkg/security/utils/mount.go]: Fixed

File: cmd/security-agent/app/runtime_unsupported.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (lebauce) [reply cmd/security-agent/app/runtime_unsupported.go]: Changed to an error

File: releasenotes/notes/consistent-config-secret-decryption-f24dd89f5c485f88.yaml (hunk 5 above)
Comment (lebauce) [reply releasenotes/notes/consistent-config-secret-decryption-f24dd89f5c485f88.yaml]: Added changelog and removed label

File: pkg/config/config_template.yaml (hunk 2 above)
Comment (lebauce) [reply pkg/config/config_template.yaml]: Changed to `runtime_security_config.policies.dir`

File: cmd/agent/dist/conf.d/runtime_security_agent.d/default.policy [hunk 15]
```diff
@@ -0,0 +1,53 @@
+- rule:
```
Comment (clamoriniere) [inline cmd/agent/dist/conf.d/runtime_security_agent.d/default.policy:1]: my 2 cents, 
I think it can be good to version this file by adding a kind of header to the file structure. because if never this file can be provided by the user, it will ease to support backward and forrward compatibility between the agent and the file format.

```yaml
version: 1.0.0
rules:
  - id: credential_modified
    description: credential files modified using unknown tool
    expression: >-
      (open.filename == "/etc/shadow" || open.filename == "/etc/gshadow") &&
      process.name not in ["vipw", "vigr"]
    tags:
      mitre: T1003
```

File: cmd/system-probe/modules/security_runtime.go [hunk 16]
```diff
@@ -0,0 +1,22 @@
+package modules
```
Comment (clamoriniere) [inline cmd/system-probe/modules/security_runtime.go:1]: missing code header

Comment (lebauce) [reply cmd/system-probe/modules/security_runtime.go:1]: Are you talking about the license header or about Go build constraints

File: cmd/agent/dist/conf.d/runtime_security_agent.d/default.policy (hunk 15 above)
Comment (lebauce) [reply cmd/agent/dist/conf.d/runtime_security_agent.d/default.policy]: Good point. Updated

File: cmd/system-probe/modules/security_runtime.go (hunk 16 above)
Comment (lebauce) [reply cmd/system-probe/modules/security_runtime.go:6]: Done

File: go.mod [hunk 17]
```diff
@@ -77,6 +77,7 @@ require (
 	github.com/containerd/typeurl v1.0.0
 	github.com/coreos/go-semver v0.3.0
 	github.com/coreos/go-systemd v0.0.0-20190620071333-e64a0ec8b42a
+	github.com/davecgh/go-spew v1.1.1
```
Comment (xornivore) [inline go.mod:80]: It doesn't look like we are using it anymore in this PR, do we need it?

File: pkg/config/config_template.yaml [hunk 18]
```diff
@@ -873,6 +873,42 @@ api_key:
...
+  # socket: /opt/datadog-agent/run/runtime-security.sock
+
+  ## @param policies - custom object - optional
+  ## Policy files
+  # policies:
+
+    ## @param dir - string - folder from where the policies will be loaded
+    ## Path from where the policy files will be loaded
+    #
+    dir: /etc/datadog-agent/runtime_security_agent.d
```
Comment (xornivore) [inline pkg/config/config_template.yaml:900]: Nit - maybe use dashes in paths - other paths seem to follow that convention?

File: pkg/security/ebpf/byteorder.go [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+package ebpf
+
+import (
+	"encoding/binary"
+	"unsafe"
+)
+
+// GetHostByteOrder guesses the hosts byte order
+func GetHostByteOrder() binary.ByteOrder {
```
Comment (xornivore) [inline pkg/security/ebpf/byteorder.go:14]: Should this be in `pkg/util` instead?

File: pkg/security/ebpf/c/chmod.h [hunk 20]
```diff
@@ -0,0 +1,97 @@
+#ifndef _CHMOD_H_
```
Comment (xornivore) [inline pkg/security/ebpf/c/chmod.h:1]: Should probably have comment headers on these files as well?

File: pkg/security/ebpf/table.go (hunk 4 above)
Comment (xornivore) [reply pkg/security/ebpf/table.go:106]: Yes, but do you need to know about this type `StringTableItem` outside of this package, you just use it via the interface? Why export this type and other types implementing the `TableItem`. E.g. here: https://github.com/DataDog/datadog-agent/pull/6014/files#diff-55f7608a287af8fc4223cf27806e86b3R186. Am I missing some other use case?

File: pkg/config/config.go [hunk 21]
```diff
@@ -733,6 +733,14 @@ func InitConfig(config Config) {
 	config.BindEnvAndSetDefault("compliance_config.dir", "/etc/datadog-agent/compliance.d")
 	config.BindEnvAndSetDefault("compliance_config.cmd_port", 5010)
 
+	// Datadog security agent (runtime)
+	config.BindEnvAndSetDefault("runtime_security_config.enabled", false)
+	config.BindEnvAndSetDefault("runtime_security_config.debug", false)
```
Comment (xornivore) [inline pkg/config/config.go:738]: This is missing from the example config.

File: pkg/config/config.go [hunk 22]
```diff
@@ -733,6 +733,14 @@ func InitConfig(config Config) {
...
+	config.BindEnvAndSetDefault("runtime_security_config.policies.dir", "")
+	config.BindEnvAndSetDefault("runtime_security_config.socket", "")
+	config.BindEnvAndSetDefault("runtime_security_config.enable_kernel_filters", true)
```
Comment (xornivore) [inline pkg/config/config.go:741]: This also is missing from the example config.

File: pkg/security/ebpf/byteorder.go (hunk 19 above)
Comment (lebauce) [reply pkg/security/ebpf/byteorder.go:14]: This is related to eBPF so it's probably ok here

File: pkg/security/ebpf/table.go (hunk 4 above)
Comment (lebauce) [reply pkg/security/ebpf/table.go:106]: When you look at the generated doc, it's clearer to see the type and its description and not just the constructor.

File: go.mod (hunk 17 above)
Comment (lebauce) [reply go.mod:80]: It is used in `pkg/security/secl/generators/accessors/accessors.go`

File: pkg/security/ebpf/c/chmod.h (hunk 20 above)
Comment (lebauce) [reply pkg/security/ebpf/c/chmod.h:1]: Obviously, we can't have the usual headers on top of these :-) So I suggest we just leave it as it

File: pkg/security/agent/agent.go [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+// NewRuntimeSecurityAgent instantiates a new RuntimeSecurityAgent
+func NewRuntimeSecurityAgent() (*RuntimeSecurityAgent, error) {
+	hostname, err := util.GetHostname()
+	if err != nil {
+		return nil, err
+	}
+
+	path := fmt.Sprintf("unix://%s", coreconfig.Datadog.GetString("runtime_security_config.socket"))
```
[the remark this answers is not in the document]
Comment (lebauce) [reply pkg/security/agent/agent.go:41]: Changed to returning an error if it's not set. If it doesn't exist, the `Dial` will return an error

File: pkg/security/config/config.go [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+	SocketPath          string
+	SyscallMonitor      bool
+}
+
+// NewConfig returns a new Config object
+func NewConfig() (*Config, error) {
+	c := &Config{
+		Enabled:             aconfig.Datadog.GetBool("runtime_security_config.enabled"),
+		Debug:               aconfig.Datadog.GetBool("runtime_security_config.debug"),
+		EnableKernelFilters: aconfig.Datadog.GetBool("runtime_security_config.enable_kernel_filters"),
```
[the remark this answers is not in the document]
Comment (lebauce) [reply pkg/security/config/config.go:34]: Added documentation to configuration template

File: pkg/config/config.go (hunk 21 above)
Comment (lebauce) [reply pkg/config/config.go:738]: Added documentation to configuration template

File: pkg/config/config.go (hunk 22 above)
Comment (lebauce) [reply pkg/config/config.go:741]: Added documentation to configuration template

File: pkg/config/config_template.yaml (hunk 18 above)
Comment (lebauce) [reply pkg/config/config_template.yaml:905]: We followed the convention of folders in `cmd/agent/dist/conf.d`. Should we use the dash convention ?

Comment (xornivore) [reply pkg/config/config_template.yaml:905]: We already have `-` in `datadog-agent` in this path, and the socket path uses `-`.

Comment (xornivore) [reply pkg/config/config_template.yaml:905]: How about `/etc/datadog-agent/runtime-security.d`? Can we also set this as default in the config env binds?

File: pkg/security/agent/agent.go [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+		return nil, err
+	}
+
+	socketPath := coreconfig.Datadog.GetString("runtime_security_config.socket")
+	if socketPath == "" {
+		return nil, errors.New("runtime_security_config.socket must be set")
+	}
+
+	path := "unix://" + socketPath
+	conn, err := grpc.Dial(path, grpc.WithInsecure())
```
Comment (xornivore) [inline pkg/security/agent/agent.go:47]: Could you please add a reasonable retrier here and possibly increase the timeout option here (would be nice to add a test for that too). There is a race right now - it fail to dial if the sysprobe container didn't bring up the socket yet (running in separate containers). To confirm the socket actually exists:
```
# inside the security-agent container:
root@datadog-monitoring-8nvff:/# ls /var/run/sysprobe/
runtime-security.sock  sysprobe.sock
```

Comment (lebauce) [reply pkg/security/agent/agent.go:47]: There already is a retry mechanism so that we don't fail if system-probe hasn't created the socket yet (agent.go:88)

File: pkg/config/config_template.yaml (hunk 18 above)
Comment (lebauce) [reply pkg/config/config_template.yaml]: Changed to /etc/datadog-agent/runtime-security.d and updated default

File: pkg/security/agent/agent.go (hunk 25 above)
Comment (xornivore) [reply pkg/security/agent/agent.go:47]: Hm.. maybe it actually does connect after retries, just there isn't an indication in the logs it succeeded eventually. Maybe add a log saying "successfully connected" to socket?

Comment (safchain) [reply pkg/security/agent/agent.go:47]: I added a log when there is a connection success after each connection fail

File: pkg/config/config.go [hunk 26]
```diff
@@ -736,8 +736,8 @@ func InitConfig(config Config) {
 	// Datadog security agent (runtime)
 	config.BindEnvAndSetDefault("runtime_security_config.enabled", false)
 	config.BindEnvAndSetDefault("runtime_security_config.debug", false)
-	config.BindEnvAndSetDefault("runtime_security_config.policies.dir", "")
-	config.BindEnvAndSetDefault("runtime_security_config.socket", "")
+	config.BindEnvAndSetDefault("runtime_security_config.policies.dir", "/etc/datadog-agent/runtime-security.d")
+	config.BindEnvAndSetDefault("runtime_security_config.socket", "/opt/datadog-agent/run/runtime-security.sock")
```
Comment (xornivore) [inline pkg/config/config.go:740]: We should probably change to `/var/run/sysprobe/runtime-security.sock` to avoid an extra mount from system probe.

File: pkg/config/config_template.yaml [hunk 27]
```diff
@@ -899,10 +899,10 @@ api_key:
   ## Policy files
   # policies:
 
-    ## @param dir - string - folder from where the policies will be loaded
+    ## @param dir - string - default: /etc/datadog-agent/runtime-security.d
     ## Path from where the policy files will be loaded
     #
-    dir: /etc/datadog-agent/runtime_security_agent.d
+    # dir: /etc/datadog-agent/runtime-security.d
```
Comment (xornivore) [inline pkg/config/config_template.yaml:905]: When this dir is not initialized - system probe currently fails to start. Do we create this dir in the docker image?

File: pkg/config/config.go (hunk 26 above)
Comment (safchain) [reply pkg/config/config.go:740]: where `/var/run/sysprobe/` is specified ? We saw `/opt/datadog-agent/run/sysprobe.sock` here https://github.com/DataDog/datadog-agent/blob/614a67138f54ed88f340897088cb3293570f7798/Dockerfiles/manifests/all-containers/system-probe-configmap.yaml#L13

Comment (safchain) [reply pkg/config/config.go:740]: or here : https://github.com/DataDog/datadog-agent/blob/master/pkg/config/config_template.yaml#L869

Comment (xornivore) [reply pkg/config/config.go:740]: Yep, you are totally right, please keep this like you configured. The `/var/run/sysprobe/runtime-security.sock` is specific to the public helm chart so I will use that default there.

File: pkg/config/config_template.yaml (hunk 27 above)
Comment (safchain) [reply pkg/config/config_template.yaml:905]: We added the folder creation during the package creation

**Review by xornivore — APPROVED** [review]
🚀

File: releasenotes/notes/runtime-security-agent-fe14b8fab5b6ee9e.yaml [hunk 28]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    The new Runtime Security Agent collects file integrity monitoring events.
```
Comment (olivielpeau) [inline releasenotes/notes/runtime-security-agent-fe14b8fab5b6ee9e.yaml:4]: let's mention here:

* that it's disabled by default
* that it only supports Linux (for now?)
* whether it'll be in beta or stable (the release note can be updated later if that's not something that can be publicly documented yet)
* how users can enable it (or a link to public docs? Again the release note can be updated later if that's not public yet)

File: pkg/config/render_config.go [hunk 29]
```diff
@@ -41,6 +41,7 @@ type context struct {
 	TraceAgent        bool
 	ClusterChecks     bool
 	CloudFoundryBBS   bool
+	SecurityModule    bool
```
Comment (olivielpeau) [inline pkg/config/render_config.go:44]: this is false by default, so won't be included unless you set it to true for some agent build types in `mkContext`

File: omnibus/config/software/datadog-agent.rb [hunk 30]
```diff
@@ -78,6 +78,7 @@
     conf_dir = "#{install_dir}/etc/datadog-agent"
   end
   mkdir conf_dir
+  mkdir "#{conf_dir}/runtime-security.d"
```
Comment (olivielpeau) [inline omnibus/config/software/datadog-agent.rb:81]: could make sense to only include on Linux, since it's not available on other platforms

**Review by olivielpeau — COMMENTED** [review]
Very interesting stuff! Left a few comments/questions, I reviewed most files except `pkg/ebpf` and `pkg/security` (where I don't think my review will bring much value)

File: cmd/security-agent/app/compliance.go [hunk 31]
```diff
@@ -64,16 +77,118 @@ func eventRun(cmd *cobra.Command, args []string) error {
...
+	runner := runner.NewRunner()
+	stopper.Add(runner)
+
+	scheduler := scheduler.NewScheduler(runner.GetChan())
+	runner.SetScheduler(scheduler)
+
+	checkInterval := coreconfig.Datadog.GetDuration("compliance_config.check_interval")
+	configDir := coreconfig.Datadog.GetString("compliance_config.dir")
+
+	hostname, err := util.GetHostname()
```
[the remark this answers is not in the document]
Comment (truthbk) [reply cmd/security-agent/app/compliance.go:153]: The gRPC endpoint here is available and ready for use. No authentication is necessary to use this endpoint and is available at `pb.Agent.GetHostname` [here](https://github.com/DataDog/datadog-agent/blob/master/cmd/agent/api/pb/api.proto#L10-L14). Also available at https://localhost:5001/v1/grpc/host, via the grpc-gateway, but the protobuf/gRPC way would be the preferred approach.

Let me know if any help is needed with this. Something to keep in mind:
1) be resilient to endpoint failures, although we can set service dependencies on systemd it might not be enough to guarantee the API is up and fully ready by the time the security agent queries for the canonical hostname, so I recommend adding some retry mechanism.
2) if we fail to collect the hostname via the API, then we can fallback to the legacy way of grabbing the hostname via a shellout.

File: omnibus/config/software/datadog-agent.rb (hunk 30 above)
Comment (truthbk) [reply omnibus/config/software/datadog-agent.rb:81]: +1

Comment (lebauce) [reply omnibus/config/software/datadog-agent.rb:81]: Indeed. Fixed

File: pkg/config/render_config.go (hunk 29 above)
Comment (lebauce) [reply pkg/config/render_config.go:44]: Good catch. Fixed

File: cmd/security-agent/app/compliance.go [hunk 32]
```diff
@@ -64,16 +77,118 @@ func eventRun(cmd *cobra.Command, args []string) error {
...
+	endpoints, err := config.BuildEndpoints(httpConnectivity)
+	if err != nil {
+		return nil, log.Errorf("Invalid endpoints: %v", err)
+	}
+
+	destinationsCtx := client.NewDestinationsContext()
+	destinationsCtx.Start()
+	stopper.Add(destinationsCtx)
+
+	health := health.RegisterLiveness("security-agent")
```
[the remark this answers is not in the document]
Comment (lebauce) [reply cmd/security-agent/app/compliance.go:110]: Changed to `compliance-security-agent`. Is it ok for you @xornivore

File: cmd/security-agent/app/compliance.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (lebauce) [reply cmd/security-agent/app/compliance.go:153]: Makes sense.
The security-agent is built using the `docker` and `kubelet` build tags so most hostname provider will work. If we later choose to forward the logs to the agent, we would not need to retrieve the hostname at all. If this is ok for you @olivielpeau and @truthbk, I suggest we left it as it until we decide whether we keep using a dedicated log client or forward the logs to the agent.

File: cmd/security-agent/app/compliance.go [hunk 33]
```diff
@@ -64,16 +77,118 @@ func eventRun(cmd *cobra.Command, args []string) error {
...
 	return nil
 }
+
+func newComplianceReporter(stopper restart.Stopper, sourceName, sourceType string) (event.Reporter, error) {
+	httpConnectivity := config.HTTPConnectivityFailure
+	if endpoints, err := config.BuildHTTPEndpoints(); err == nil {
+		httpConnectivity = http.CheckConnectivity(endpoints.Main)
+	}
+
+	endpoints, err := config.BuildEndpoints(httpConnectivity)
```
[the remark this answers is not in the document]
Comment (lebauce) [reply cmd/security-agent/app/compliance.go:101]: You're right.
1. If I remember correctly, this design choice was done because logs collection may be disabled on the agent. Any other idea @xornivore ?
2. I removed the DDClient to be more consistent
3. I updated the PR so that config.Endpoints and client.DestinationsContext are shared.

I also added a comment in config_template.yaml to explain to explain that configuration settings required to connect to DD are fetch from the `logs_config` section.

File: cmd/security-agent/app/compliance.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (olivielpeau) [reply cmd/security-agent/app/compliance.go:153]: ok, I quickly checked the implementation of `util.GetHostname` and at the moment the `docker` and `kubelet` build tags seem to be the only ones that can affect it.

Your plan sounds good to me, but please make sure this is tracked on your end in case additional build tags in the future affect the hostname resolution logic. Hostname inconsistencies can have billing implications. Also, can you add a `FIXME` comment inline?

File: releasenotes/notes/runtime-security-agent-fe14b8fab5b6ee9e.yaml (hunk 28 above)
Comment (lebauce) [reply releasenotes/notes/runtime-security-agent-fe14b8fab5b6ee9e.yaml:4]: Updated the release note. Regarding whether it's beta or stable and the documentation, we'll update it later it's it ok as all details are not settled yet.

File: cmd/security-agent/app/compliance.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (lebauce) [reply cmd/security-agent/app/compliance.go]: Ok. Also added a FIXME

File: cmd/agent/dist/conf.d/runtime_security_agent.d/default.policy (hunk 15 above)
Comment (xornivore) [reply cmd/agent/dist/conf.d/runtime_security_agent.d/default.policy]: Is there code to check the policy version is compatible with the agent?

File: cmd/security-agent/app/compliance.go (hunk 32 above)
[the remark this answers is not in the document]
Comment (xornivore) [reply cmd/security-agent/app/compliance.go]: I think it should still be "security-agent". Compliance agent is a misnomer, we shouldn't have it in anything user facing.

File: cmd/security-agent/app/compliance.go (hunk 33 above)
[the remark this answers is not in the document]
Comment (xornivore) [reply cmd/security-agent/app/compliance.go]: Yes, so logs collection being disabled is one concern. Another consideration is that the logs intake will eventually be different for regular logs and the security agent events.

File: cmd/agent/dist/conf.d/runtime_security_agent.d/default.policy (hunk 15 above)
Comment (lebauce) [reply cmd/agent/dist/conf.d/runtime_security_agent.d/default.policy]: Not yet, we'll add this check when required

File: cmd/security-agent/app/compliance.go (hunk 32 above)
[the remark this answers is not in the document]
Comment (lebauce) [reply cmd/security-agent/app/compliance.go]: As discussed, changed to `compliance`

File: cmd/security-agent/app/compliance.go (hunk 33 above)
[the remark this answers is not in the document]
Comment (olivielpeau) [reply cmd/security-agent/app/compliance.go]: Ok, thanks for the context

**Review by olivielpeau — APPROVED** [review]

**Review by apigirl — COMMENTED** [review]
docs team has reviewed. thanks for letting us know! we'll reach out to coordinate docs for dash

**Review by apigirl — APPROVED** [review]
pkg/config/config_template.yaml looks good from docs team
